src/uwtools/template.py

```diff
@@ -0,0 +1,172 @@
+
+# (C) Copyright 2020-2022 UCAR
+#
+# This software is licensed under the terms of the Apache Licence Version 2.0
+# which can be obtained at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Part of this software is developed by the Joint Center for Satellite Data Assimilation (JCSDA) together with its partners.
+#
+# (C) Copyright 2022-2022 NOAA
+
+import re
+import os
+import copy
+from collections import namedtuple
+from collections.abc import Sequence
+
+class TemplateConstants:
+    DOLLAR_CURLY_BRACE = '${}'
+    DOLLAR_PARENTHESES = '$()'
+    DOUBLE_CURLY_BRACES = '{{}}'
+    AT_SQUARE_BRACES = '@[]'
+    AT_ANGLE_BRACKETS = '@<>'
+
+    SubPair = namedtuple('SubPair', ['regex', 'slice'])
+
+
+class Template:
+
+    """
+        Utility for substituting variables in a template. The template can be the contents of a whole file
+        as a string (substitute_string) of in a complex dictionary (substitute_structure).
+        substitutions defines different type of variables with a regex and a slice:
+        - the regex is supposed to find the whole variable, e.g, $(variable)
+        - the slice indicate how to slice the value returned by the regex to have the variable name, in the
+          case of $(variable), the slice is 2, -1 to remove $( and ).
+        You can easily add new type of variables following those rules.
+
+        Please note that the regexes allow for at least one nested variable and the code is able to handle it.
+        It means that $($(variable)) will be processed correctly but the substitutions will need more than one
+        pass.
+
+        If you have a file that is deeper than just a simple dictionary of has lists in it, you can use the method
+        build_index to create a dictionary that will have all the options from deeper levels (list, dicts).
+        You can then pass index.get as an argument to any method you use.
+        If you use substitute_with_dependencies, this is done automatically.
+    """
+
+    substitutions = {
+        TemplateConstants.DOLLAR_CURLY_BRACE: TemplateConstants.SubPair(re.compile('\${.*?}+'), slice(2, -1)),
+        TemplateConstants.DOLLAR_PARENTHESES: TemplateConstants.SubPair(re.compile('\$\(.*?\)+'), slice(2, -1)),
+        TemplateConstants.DOUBLE_CURLY_BRACES: TemplateConstants.SubPair(re.compile('{{.*?}}+'), slice(2, -2)),
+        TemplateConstants.AT_SQUARE_BRACES: TemplateConstants.SubPair(re.compile('@\[.*?\]+'), slice(2, -1)),
+        TemplateConstants.AT_ANGLE_BRACKETS: TemplateConstants.SubPair(re.compile('@\<.*?\>+'), slice(2, -1))
+    }
+
+    def is_single_type_or_string(s):
+        if isinstance(s, str):
+            return True
+        try:
+            len(s)
+        except TypeError:
+            return True
+        else:
+            return False    
+
+    @classmethod
```

Comment:
IMHO, would be better served as module functions versus class based functions.  Would offer project wide re-usability, for both methods and constants, without having to invoke/reference the ```Template``` class.